### ISSUE TITLE
Android App

### ISSUE THREAD

**Issue by abhijithvijayan** [thedevs-network/kutt#165]
Is it be useful to build an android app for Kutt?

**Comment by trgwii** [comment]
Not sure, when was the last time you shortened a link on your phone?

**Comment by abhijithvijayan** [comment]
@trgwii I use bit.ly app all the time. Click the long url and it says bit.ly and it will automatically shorten and copy to clipboard in the background.

**Comment by poeti8** [comment]
I liked the idea of having the kutt button alongside copy.

**Comment by abhijithvijayan** [comment]
@poeti8 I can work on the app.

**Comment by Yash-Garg** [comment]
> I liked the idea of having the kutt button alongside copy.

Me too, i think it'sa good idea to have one!

**Comment by abhijithvijayan** [comment]
@poeti8 But what about authentication? 

I was thinking of forwarding them to the kutt endpoint via the browser and making them copy the generated api key and paste it in the app.

If there was an endpoint that returned the api key(if exist) on POST ing the id and password, it could've been easier.

Does it affect the security if such an endpoint exist ?

**Comment by poeti8** [comment]
How does Bitly do it? I prefer to not create such an endpoint.

**Comment by abhijithvijayan** [comment]
@poeti8 They use authentication

**Comment by poeti8** [comment]
@abhijithvijayan What's the process? Everytime user wants to shorten the URL or only once?

**Comment by abhijithvijayan** [comment]
@poeti8 After authentication, as from bit.ly, anyone can copy the long url and click on the bit.ly icon next to cut copy and it will auto shorten and copy the shortened url to clipboard in the background itself.

**Comment by abhijithvijayan** [comment]
also opening the app provides option to shorten new urls.

All the shortened urls created with that account will list on the app.

**Comment by poeti8** [comment]
@abhijithvijayan Ok, what's stopping us from implementing this?

**Comment by abhijithvijayan** [comment]
@poeti8 The authentication and also the listing requires returning all the created urls back to the app via some api call.

**Comment by poeti8** [comment]
We already have an endpoint to return the list of URLs.
Also logged in users can generate an API or see the existing one.

What else needs to be done? You need to be specific.

**Comment by abhijithvijayan** [comment]
@poeti8 I guess as CORS is enabled, I can send POST request to the /auth/something endpoint for the login.

Maybe that's enough. 

Let me try with an API client for testing.

**Comment by abhijithvijayan** [comment]
@poeti8 Why is the login not successful? It works from browser but not from api client?

![Insomnia_2019-06-10_20-43-10](https://user-images.githubusercontent.com/34790378/59205430-89007100-8bc0-11e9-8e65-13604cb51602.png)

**Comment by poeti8** [comment]
@abhijithvijayan send as JSON body.

**Comment by abhijithvijayan** [comment]
> @abhijithvijayan send as JSON body.

How should I attach the token that was received has to be attached for the rest of the end points?

**Comment by abhijithvijayan** [comment]
The app will work like this
For login
sends a POST request to `/api/auth/login`
if success, 
- gets api key (GET to `/api/auth/usersetting`)
   - if not exist 
           generate one (POST to `/api/auth/generateapikey`)
- also gets the url history (GET to `/api/url/geturls`)

On submission of long urls, it might be better to give the custom url option after once the url is shortened (like edit url option)

**Comment by poeti8** [comment]
> The only thing I am missing here is getting api key(or generating it) with the login token obtained after loggin in.

When login you'll get a token, when got the token now you can call the other endpoints with the API key in the header of request: `{ headers: { 'X-API-KEY': 'samplekey' } }`

**Comment by abhijithvijayan** [comment]
@poeti8 what if the api key is not generated?

**Comment by poeti8** [comment]
> Edit: The token I get after logging in, is to be send in Authorization Header as a Bearer Token, for the subsequent requests, right? or is it in the query parameter?

This is all in the Kutt readme. Send it as `X-API-TOKEN` header.

**Comment by xianx** [comment]
What is the progress of kutt android app? A react native app can be build.

**Comment by abhijithvijayan** [comment]
@xianx Yeah I was thinking of the same. Currently it is not started yet as I was busy with other projects.

Will start working on it soon. Thanks.

**Comment by NewRedsquare** [comment]
Any news ?

**Comment by abhijithvijayan** [comment]
I havent had a chance to work on it. Sorry.

**Comment by atommarvel** [comment]
Just found Kutt which is exactly what I was looking for :)

I've started working on the android app. Feel free to follow progress here: https://github.com/atommarvel/kutt-android

**Comment by NewRedsquare** [comment]
@atommarvel congrats and thank you a LOT for beginning this work !

I switched to another selfhosted url shortener only due to the lack of android app... i'll probably go back to kutt then

**Comment by atommarvel** [comment]
@abhijithvijayan / @MKRhere or any Kutt owner!

Question semi-related to this issue!
 
For the android app icon would yall prefer it to:
* use the Kutt logo
* incorporate the Kutt logo
* not use the Kutt logo at all
* you don't care

Also, feel free to assign this issue to me if you want :)

**Comment by abhijithvijayan** [comment]
@poeti8 ^^

**Comment by MKRhere** [comment]
You can use the Kutt logo, but have a clear indication that this is not from Kutt developers (or mention "unofficial" somewhere). Repo description, marketplace description and "about app" section (if any) are good places.

**Comment by atommarvel** [comment]
Sounds great. I've updated the repo about it being unofficial.

I also have it prominently mentioned in settings where the user sets the api key:

<img src="https://user-images.githubusercontent.com/3476374/119678646-8b335880-bdf4-11eb-835a-a5c5b6adb09a.png" height="500"/>

**Comment by MKRhere** [comment]
@atommarvel I like the idea of "X for Kutt"; clear that it's independent, benefits from SEO, and doesn't conflict with other implementations.

 Good job on the app so far!

**Comment by atommarvel** [comment]
Cool. I'll go with "Shears for Kutt".

Thinking about it more, I realize that the "K" logo doesn't make a ton of sense for an app named Shears. 

Would it be okay to modify the logo in order to clearly communicate an association AND a distinction?


Current draft:
<img width="663" alt="Screen Shot 2021-05-26 at 3 00 15 PM" src="https://user-images.githubusercontent.com/3476374/119716418-f80d1980-be19-11eb-99f5-235acf002336.png">

**Comment by MKRhere** [comment]
@atommarvel Looks good 🙂

**Comment by poeti8** [comment]
Thanks @atommarvel , let me know when it's done and I'll add to the README.

**Comment by Sulphuric6297** [comment]
What ever came of this? An android app would be fantastic.

**Comment by trgwii** [comment]
Inactivity
